src/main/java/bio/terra/service/filedata/flight/delete/FileDeleteFlight.java

```diff
@@ -44,10 +50,10 @@ public FileDeleteFlight(FlightMap inputParameters, Object applicationContext, Us
         // 4. Delete the directory entry
         // This flight updates GCS and firestore in exactly the reverse order of create, so no new
         // data structure states are introduced by this flight.
-        addStep(new DeleteFileLookupStep(fileDao, fileId, dataset, dependencyDao));
-        addStep(new DeleteFileMetadataStep(fileDao, fileId, dataset));
+        addStep(new DeleteFileLookupStep(fileDao, fileId, dataset, dependencyDao), fileSystemRetry);
```

Comment:
I wonder if there's a better way to create a relationship between steps and retryability here. It looks like a retry rule is a per-step setting -- and that makes sense to me, since only a step can know if it could be retried -- but it isn't passed to the step constructor. Is there a reason why? Another way to think about this is, does it make sense to use a retryable step like `DeleteFileLookupStep` without retry, or should the caller be forced to provide a retry for it?

Also the same rule is used for all retryable steps. If a step could itself claim that it's retryable, and if the same retry rule is used for all steps in a flight, then the rule could be set on the flight instead of having it be per-step. As it is now, the retry rule is mutable and is shared between multiple objects. Having a single reference to a mutable object makes code behavior clearer, IMO.

Reply:
All steps have a retry rule. It defaults to a no-retry implementation.

All steps need to be retryable (idempotent): the DR Manager may crash. On restart, the flight is reconstituted from the database state and the step is re-run.

Different steps in a flight need different retry logic. For example, if a SAM operation times out, we probably want a fairly long, regular interval retry. Whereas the concurrency issue in the filesystem code wants to spread threads randomly within a finite timeframe. So the step is definitely the right unit of retry definition.

The DR code is not mature in its use of Stairway retry. There are lots of cases of cloud resource unavailability that should be retried, but currently are not.